src/ray/core_worker/object_interface.cc

```diff
@@ -12,14 +12,25 @@ Status CoreWorkerObjectInterface::Put(const Buffer &buffer, ObjectID *object_id)
   ObjectID put_id = ObjectID::ForPut(core_worker_.worker_context_.GetCurrentTaskID(),
                                      core_worker_.worker_context_.GetNextPutIndex());
   *object_id = put_id;
+  return Put(buffer, put_id);
+}
 
-  auto plasma_id = put_id.ToPlasmaId();
+Status CoreWorkerObjectInterface::Put(const Buffer &buffer, const ObjectID &object_id) {
+  auto plasma_id = object_id.ToPlasmaId();
   std::shared_ptr<arrow::Buffer> data;
-  RAY_ARROW_RETURN_NOT_OK(
-      core_worker_.store_client_.Create(plasma_id, buffer.Size(), nullptr, 0, &data));
+  {
+    std::unique_lock<std::mutex> guard(core_worker_.store_client_mutex_);
```

Comment:
No need to do anything about this in this PR, but I wonder if `store_client_mutex_` should actually be part of the plasma client (in the Arrow codebase) instead of part of the `CoreWorkerObjectInterface`. The reason is that in Python, when we call `x = ray.get(object_id)`. If `x` is a numpy array, then when `x` goes out of scope in Python, a call to `store_client_.Release(object_id)` will be made. This happens outside of the `CoreWorkerObjectInterface`, and so we don't have a good way to lock it with the mutex here.

Does that make sense? The relevant call to `Release` is in the `__dealloc__` method in https://github.com/apache/arrow/blob/997226a9263430bc0422189180bc2551aed1f63d/python/pyarrow/_plasma.pyx#L215-L255, which gets called in the numpy array destructor.

cc @pcmoritz

Reply:
Yeah, I agree with this!

Reply:
yes, this is a good point. I've filed an [issue](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-5533) for this (and self assigned)